### DISCUSSION TITLE
When syntax highlighting is enabled, Asian text become dim

### DISCUSSION THREAD

**Discussion by HubKing** [tsujan/FeatherPad#795] [Q&A] [answered]
When syntax highlighting is enabled, and the current document is a new one, shouldn't s.h. not work? But it seems like in this case, if I type Japanese or Korean, text becomes dim. Is this an intended feature, or a bug of FeatherPad or Fcitx5 (the IME)?

[Screencast_20240504_012007.webm](https://github.com/tsujan/FeatherPad/assets/10900489/5eeb51b2-5293-481c-9f11-1e881d976dd1)

**Comment by tsujan** [comment]
> When syntax highlighting is enabled, and the current document is a new one, shouldn't s.h. not work?

The default syntax highlighting is that of URL, and it works with a new document. Type `http://a` to see it.

I don't know how Asian texts work, but your video shows that the text is OK in the end. The reason it's gray at first may be that highlighting hasn't finished yet, waiting for the final text.

**Reply by HubKing** [reply]
I don't know why all Asian text's input in composition should match the URL pattern. Seems like a bug.  Also, it seems that it happens with all the text in the line before the current composition from line two. This is very distracting.

[Screencast_20240504_025906.webm](https://github.com/tsujan/FeatherPad/assets/10900489/bb0f7b5a-4308-488f-b142-2f2203953d57)

If fixing this is difficult, I wish I could disable URL highlighting. Actually, URL lighting has been in general problematic for me because I could accidentally click the URL when trying to select the text.

**Reply by HubKing** [reply]
Seems like it's caused by "Show whitespace" and "Also show line ends". It's just that disabling syntax highlighting disables those whitespaces, too.

**Reply by tsujan** [reply, marked as answer]
> I don't know why all Asian text's input in composition should match the URL pattern

They don't need to match; the highlighter should wait to see what the final text is.

> Seems like it's caused by "Show whitespace" and "Also show line ends".

Yes. But it isn't a bug. It's how whitespaces are made gray: the whole text is gray before highlighting, and the real text gets real colors after it, leaving the whitespaces.

**Reply by HubKing** [reply]
Maybe it's not a bug technically, but in users' perspective, I think this shouldn't be a desirable behaviour. I guess I could just disable whitespace options.

**Reply by tsujan** [reply]
> I could accidentally click the URL when trying to select the text.

URLs aren't launched when clicked; `Ctrl` should be pressed while clicking.

**Reply by tsujan** [reply]
> but in users' perspective,

I understand. Why Fcitx5 interrupts text processing, I have no clue.

**Comment by tsujan** [comment]
> Seems like a bug

I agree, but I doubt that it can be in FeatherPad. A coder who knows Asian texts in addition to Qt coding may have a solution — typing what's in your videos is beyond me ;)

**Reply by HubKing** [reply]
I don't know QT well, but maybe Featherpad is not accurately getting Asian word locations. If you look closely, KWrite is showing the "space dot" in correct positions, between the Korean words, whereas Featherpad's space dot is overlapping with the words. KWrite doesn't have such a problem, so maybe you could find some hints in KWrite's source code.

![image](https://github.com/tsujan/FeatherPad/assets/10900489/7545d15a-8800-4cd0-a58c-1edf995c8d64)

**Reply by tsujan** [reply]
>  maybe Featherpad is not accurately getting Asian word locations

FeatherPad doesn't get Asian words; Qt does.

> KWrite is showing the "space dot" in correct positions....

KWrite uses KDE libraries for text processing, while FeatherPad uses Qt. Using KDE libraries in FeatherPad is not an option.

**Reply by tsujan** [reply]
OK, this is definitely a Qt bug.

In FeatherPad's code, I left the text to be painted by Qt completely, instructed Qt to show whitespaces, copy-pasted a Korean text from the Internet (whose meaning I don't know), and added a single space to its beginning. This is what happened:

![korean](https://github.com/tsujan/FeatherPad/assets/981076/f8f12dff-97c6-445f-b7ff-a1864fb89f4f)

As you see, the whitspace is positioned incorrectly.

I can't report this to Qt because I know nothing of Asian languages; you can't report it either because they need a sample code (please note that I don't have a sample code — I just made FeatherPad's code leave it to Qt). Unless it's already reported, Qt devs don't seem to know about it (apparently, there's no one among them with a knowledge of Asian languages).

The problem might have been fixed in a KDE library by a contributor at some point.

**Reply by HubKing** [reply]
There is another text editor written in QT, although it's QT5: Notepadqq. It doesn't seem to have a problem with Korean, although it did not recognise the Japanese full-width space as whitespace. It probably isn't using KDE, so maybe you could refer to its code.

![image](https://github.com/tsujan/FeatherPad/assets/10900489/a90ea98b-83fa-40a3-8330-4003aa59aaac)

**Reply by tsujan** [reply]
> There is another text editor written in QT

All KDE apps are also written in Qt. Qt has so many classes.

FeatherPad uses [QPlainTextEdit](https://doc.qt.io/qt-6/qplaintextedit.html) because it's lightweight and fast. The problem is there. There are other classes for showing and editing text (like [QTextEdit](https://doc.qt.io/qt-6/qtextedit.html)), but they aren't for lightweight text/code editors.

**Reply by tsujan** [reply]
Quoting from https://github.com/notepadqq/notepadqq/issues/873: Notepadqq is "very RAM hungry because it's practically running a browser in which to display the text."
